### ISSUE TITLE
Alternate closure syntax

### ISSUE THREAD

**Issue by IanManske** [nushell/nushell#13750]
### Related problem

I think the concerns expressed in #4115 and #8273 are valid / worth addressing. Currently, record literals, blocks, and closures all share the same or similar syntax:
```nushell
{ foo: bar } # record
{ foo bar } # closure
loop { foo bar } # block
``` 

I think it would be helpful to add a little more clarity by introducing a new syntax for closures. This would also allow us to remove the pipe character (`|`) from the closure syntax, and perhaps make parsing a little simpler.

### Describe the solution you'd like

My tentative syntax proposal is as follows:
```nushell
# closure with no arguments
@{ foo bar }
# or
@[]{ foo bar }

# closure with arguments
@[x y]{ foo bar $x $y }
```

I'm using the `@` sign for now, but this could be another symbol if we wanted.

One nice thing about this syntax is that it is similar to how custom commands are defined:
```nushell 
def cmd [] {
}
```

Please discuss / comment! :smile:

**Comment by fdncred** [comment]
Some other potential options.
1. `{ x,y => foo bar $x $y }` # fat arrow (c# does this)
2. `{ x,y -> foo bar $x $y }` # skinny arrow
3. `(|x, y| foo bar $x $y)` # parens instead of braces
4. `closure{|x| blah $x}` # keyword
5. `delegate{|x| blah $x}` # keyword

if we're not going to use standard rust syntax for closures, my personal preference would be to stay as close to rust as possible, which I think is #3. I could also probably live with the `@` syntax above, although it may take minute to get used to it.

**Comment by weirdan** [comment]
If the argument for the alternative syntax for closures (aka anonymous functions) is that they need to be similar to custom commands (aka named functions) then you can't get more similar than just omitting the name: `def [x, y] { foo bar $x $y }`

**Comment by IanManske** [comment]
Appreciate the suggestions @fdncred, but my qualms would be:

- 1 and 2 don't solve the record/closure parsing ambiguity.
- 3 would make the use of pipes mandatory, as otherwise there is no way to tell a sub-expression apart from a closure:
  ```nushell
  (foo bar) # expression, evaluated immediately
  (|| foo bar) # closure 
  ```
  This syntax also keeps pipes, which I think we should avoid based on #4115.
- 4 and 5 also use pipes as well as requiring more keystrokes :stuck_out_tongue:
- none of them match the custom command syntax.

**Comment by IanManske** [comment]
> If the argument for the alternative syntax for closures (aka anonymous functions) is that they need to be similar to custom commands (aka named functions) then you can't get more similar than just omitting the name: `def [x, y] { foo bar $x $y }`

Oh, that is interesting. I'm thinking that a symbol would be nicer, since it would require less typing and would be visually more concise. E.g., compare:
```nushell
$table | update col @{ str upcase }
``` 
with
```nushell
$table | update col def [] { str upcase }
$table | update col (def [] { str upcase })

$table | update col def { str upcase }
$table | update col (def { str upcase })
```

**Comment by KAAtheWiseGit** [comment]
Honestly, I feel that all of the suggested syntaxes are much worse than the one we have right now.  The issues I see:

- Using a keyword is way too verbose.  Having to type it each time to do a `filter` would be annoying.
- Using parens looks weird to me, because now parens can mean both precedence _and_ scope.   I think that having parens for closures and braces for blocks would be bad, because those are two very similar concepts (execute contents with a lexical scope), which now look differently.
- Arrows are fine, but having the braces on the outside looks weird.   JavaScript simply does `(arg) => expr` or `arg => { expr }`, which looks more familiar/natural to me.
- I'm also not a fan of more sigils.  I feel that they make code readable, but this one is the most subjective.

In general, I feel that the fact that closures look similar to blocks is a good thing.  I use closures even more often than I do blocks, and I don't think that making them more complicated is a good thing.  Here are the most recent examples from my history:

```Nushell
~> $arr | each { into binary --compact } | bytes collect
~> repeat command { ls }
~> $item | update children { each { $in.id } }
```

Now that I'm looking at it, majority of my closures don't even use the arguments.  It's mostly `update`, `each`, `filter`, and custom commands.  For the built-in commands in one-liners, ease of writing is very important.  And for custom commands I actually like the fact that closures look like blocks, because I use them as blocks[^ex].

If the pipes are really such an issue, I'd propose simply replacing them, a la `{[a, b] foo bar}`.


[^ex]: `repeat command` takes a closure and repeats it, clearing the screen each time.  If I could, I'd use a block there.

**Comment by IanManske** [comment]
> Now that I'm looking at it, majority of my closures don't even use the arguments. It's mostly update, each, filter, and custom commands. For the built-in commands in one-liners, ease of writing is very important. And for custom commands I actually like the fact that closures look like blocks, because I use them as blocks.

Agreed, that is why I think the `@{ }` shorthand is important. It's the same as right now, but just with one extra character.

**Comment by KAAtheWiseGit** [comment]
I guess I just don't get what's the problem being solved here.

I don't distinguish between closures and blocks, because I see them as basically the same thing.  Closures are even implemented as a reference to a block via `BlockId`.

**Comment by NotTheDr01ds** [comment]
I get the problem, and I'm open to something that makes the difference between records and closures more distinct (visually to the developer as well as parsing).

I'm personally not opposed to changing something like `{ into datetime }` to `@{ into datetime }` if it helps.

But ...

Consider that from the developer perspective, it's already quite easy to differentiate.  @IanManske's original examples:

> ```nu
> { foo: bar } # record
> { foo bar } # closure
> loop { foo bar } # block
> ```

If I want to be more distinct, I'd just say `{|| foo bar}`.  The empty parameter set at the beginning makes it visually more clear that it is not a record.

Also perhaps consider the opposite approach -- Making *records* more distinctive. The main problem I have *visually* with records and closures is when nesting one inside the other.  Usually it's a closure that does nothing but return a record.  For instance:

```nu
[ 1 2 3 ]
| each {{ item: $in }}
```

Honestly, putting a sigil before the closure doesn't help much there:

```nu
[ 1 2 3 ]
| each @{{ item: $in }}
```

I guess `| each {|| { item: $in }}` is a little more readable, but what if you applied the sigil to the *record* instead?

```nu
[ 1 2 3 ]
| each { @{ item: $in } }
# or each {|| @{ item: $in }}
```

**Comment by NotTheDr01ds** [comment]
I'm adding some comments on the two linked issues here since they are important in this context:

* #4115: The main argument is that using the pipe symbol is confusing in a shell since pipe already has a well established meaning.
  
  Counter: Bash uses `||` as "or" and no one is confused by that.

  Alternative solution:  Just replace the pipe symbol with square brackets in the same location:

  ```nu
  [ 1 2 3 ]
  | where {[num] $num mod 2 > 0 }
  ```

  Empty params could (optionally) be expressed as `[ ]`

* #8273 advocates *requiring* an empty param set indicator, which I strongly disagree with.  I think the current approach is both convenient and quite readable when used responsibly:

  ```nu
  ls | update name { str upcase }
  ```

**Comment by IanManske** [comment]
> I guess I just don't get what's the problem being solved here.

That's fair, I should have gone into more detail / been more clear. There are several issues that I think the new syntax solves (sorry for wall of text):

1. (Mostly) unifying closure and custom command syntax
   @kubouch was possibly thinking of allowing input/output types to be specified on closures. Having closures share a similar syntax with custom commands means we don't have to create a new syntax specific only to closures just for this purpose. 
2. Parser ambiguity between records and closures
  Having closures and records share the same syntax makes the parser slightly more complicated (it requires a look-ahead). For the new (or current) parser, it would be nice to not have to handle this case.

   Additionally, empty records and empty closures share the same syntax. If the parser encounters `{}`, it is usually treated as an empty record, but not always:
   ```nushell
   let x = {} # record
  
   def cmd [x: closure] { print $x }
   cmd {} # closure

   let x: closure = {} # record + error

   $table | update col {} # record
   ```
   Of course, one should use `{|| }` instead in these cases, but I'd rather not have this small foot-gun.

3. Visual ambiguity between records and closures
   Similar to the above point, but this time concerning readability. Currently, if I see an opening `{` in nushell code, it may be the start of a record or the start of a **shorthand** closure. I have to read the contents inside the `{}` and/or read the surrounding context to figure out which one it is. The `@` makes it more clear what is a closure vs. a record at a glance. Namely, `@{` starts a closure and `{` starts a record (and blocks start with a keyword like `loop`).

   In practice, people will use `{|| }` when it makes sense for more clarity, but what is clear to one person might not be clear to another. I think it's not unreasonable to require one additional `@` symbol so that it is hopefully clear to everyone.

4. I like pipes, but...
   This is probably the most subjective issue, but I don't like how pipes are reused for the closure syntax when they are already an important piece of syntax.
   
   > Counter: Bash uses `||` as "or" and no one is confused by that.

   That is true, but the Nushell case is a little different. In bash, `||` has to appear as one token, but our current syntax allows whitespace in the closure parameters as well as a leading pipe character in pipelines. E.g.,
   ```nushell
   # a valid pipeline:
   | ls | get name

   # a valid closure:
   { | ls | get name }
   ```

   even if we disallow leading pipes, this is still possible:
   ```nushell
   {|
     ls | get name
   }
   ```

   These are quite contrived examples, but I'm not sure if this is something we want to keep in the language.

**Comment by IanManske** [comment]
> but what if you applied the sigil to the record instead?

Nushell code has the nice property that it is a superset of JSON. So, adding the sigil to the record will break this unfortunately :pensive:.

**Comment by KAAtheWiseGit** [comment]
Yeah, records are very common too, so I don't think adding a sigil would work.

My issue is that no other programming language I know of does this kind of disambiguation.  So, while I understand the ambiguities, I personally don't believe they are enough of an issue to make closures more verbose and harder to type.

Ruby syntax is a good example here.  It has "blocks":

```Ruby
[1 2 3].each do
  puts "hello"
end

[1 2 3].each do |n|
  puts n
end
```

You can pass blocks as function arguments, which, combined with a bit of sugar, allows Ruby to do this[^rb]:

```Ruby
inside_tag("p") do
  puts "Hello"
  puts "How are you?"
end
# <p>
# Hello
# How are you?
# </p>
```

I encountered it through [Crystal](https://crystal-lang.org/), which also allows braces:

```Crystal
puts File
	.read_lines("input.txt")
	.map(&.split " ")
	.map { |(card, bet)| {Hand.new(card), bet.to_i} }
	.sort_by(&.[0])
	.map_with_index(offset = 1) { |(_, bet), i| bet.to_i * i }
	.sum
```

One more thing to note here, Crystal also uses braces for tuples, but the authors of the language opted not to change the closure syntax.

This is actually a common pattern.  Another place I saw it in is Jetpack Compose.  It uses functions which take closures to create a DSL-like language:

```Kotlin
fun Row(
    ...
    // Function with no arguments and no return value
    content: () -> Unit
)

@Composable
fun ClickCounter(clicks: Int, onClick: () -> Unit) {
    Row {
        Button(onClick = onClick) {
            Text("I've been clicked $clicks times")
        }
    }
}
```

Where the braces denote lambdas with no arguments, with the argument syntax being `{ a, b -> a + b }`.


[^rb]: source: https://www.codewithjason.com/understanding-ruby-blocks/

**Comment by KAAtheWiseGit** [comment]
I just remembered one language, which does differentiate between blocks and records.  It's the famously explicit Zig, but even there the records are differentiated, not the blocks: `const pt: Point = .{ .x = 13, .y = 67 };`.

**Comment by IanManske** [comment]
> make closures more verbose and harder to type.

I don't believe the current proposal will make it harder to type? It only adds one net character. I.e.,

```nushell
{ }
{|x| } 
```

becomes

```nushell
@{ }
@[x]{ } 
```

> My issue is that no other programming language I know of does this kind of disambiguation.

Would you mind clarifying what you mean by this? Are you referring to disambiguation between records and blocks, records and closures, blocks and closures, or all of them :sweat_smile:?

For blocks and closures though, I think there is some merit in having a distinction. Closures are first class values that cannot capture outside mutable variables. Blocks, on the other hand, are not first class values, are only used as part of keyword constructs (`if`, `for`, etc.), and can use in-scope mutable variables.

**Comment by NotTheDr01ds** [comment]
> There are several issues that I think the new syntax solves

Ok, that's a bit more clear.  If there wasn't a parser benefit, then any syntax update could be *optional*.  E.g.:

> Nushell code has the nice property that it is a superset of JSON. So, adding the sigil to the record will break this unfortunately

Yeah, I was thinking that *might* could be an optional syntax, like `{||}` is today.  But if this change is going to be *required* in order to optimize the parser then that limits the options more.

**Comment by KAAtheWiseGit** [comment]
@IanManske yeah, between all three of blocks/closures/records (or, well, structs).

`@` is a single character, but it's still a modifier, which requires shift.  I use a second layer, but on a typical keyboard `@` is also on the other end of the keyboard, away from the braces.

Overall, I get that there's are technical differences between blocks and closures, but I don't think it warrants a syntax distinction.  I much prefer the style which makes closures and blocks look alike.  It's probably familiarity, because all other languages I've used do the same thing.

At the end of the day, I don't get why Nushell requires this change, when no other language I know does[^examples]


[^examples]: There's Zig, but it chose to disambiguate records.  And it also doesn't have closures.

  There's technically Python, but I really can't call `lambda` a closure with good conscience.  It does have anonymous functions, which I think could work for Nushell, but that's a different topic

**Comment by mokeyish** [comment]
> > If the argument for the alternative syntax for closures (aka anonymous functions) is that they need to be similar to custom commands (aka named functions) then you can't get more similar than just omitting the name: `def [x, y] { foo bar $x $y }`
> 
> Oh, that is interesting. I'm thinking that a symbol would be nicer, since it would require less typing and would be visually more concise. E.g., compare:
> 
> $table | update col @{ str upcase }
> with
> 
> $table | update col def [] { str upcase }
> $table | update col (def [] { str upcase })
> 
> $table | update col def { str upcase }
> $table | update col (def { str upcase })



Closure keywords should be abandoned. Python's lambda is difficult to write, has weak functions, and is very useless.

**Comment by nome** [comment]
I don't really mind whether closure syntax is `{ foo bar }` or `{|| foo bar}` or `@{foo bar}`. But *if* the latter should be implemented - may I suggest `@(foo == bar)` as a literal syntax for row conditions? This could help turn the weird special-case argument type of `where` (which can't be used in custom commands) into a properly integrated part of the language. As with cell path literals like `$.foo.bar`, the explicit `@(foo == bar)` could be shortened to `foo == bar` if the parser knows that a row-condition type is expected, so this wouldn't invalidate the most common `where` usages (although whether or not explicit `$it` should be allowed in row-condition literals is another question). Alternatively, `@(foo == bar)` could be interpreted as shorthand for `@{$in.foo == bar}`.
